programs/uxd/src/instructions/mint_with_router.rs

```diff
@@ -0,0 +1,430 @@
+        constraint = credix_lp_depository_0.load()?.credix_shares_mint == credix_lp_depository_0_shares_mint.key() @UxdError::InvalidCredixSharesMint,
+    )]
+    pub credix_lp_depository_0: AccountLoader<'info, CredixLpDepository>,
+
+    /// #16
+    #[account(mut)]
+    pub credix_lp_depository_0_collateral: Box<Account<'info, TokenAccount>>,
+
+    /// #17
+    #[account(mut)]
+    pub credix_lp_depository_0_shares: Box<Account<'info, TokenAccount>>,
+
+    /// #18
+    #[account(
+        owner = credix_client::ID,
+        seeds = [credix_lp_depository_0_global_market_state.key().as_ref(), credix_lp_depository_0.key().as_ref(), CREDIX_LP_EXTERNAL_PASS_NAMESPACE],
+        bump,
+        seeds::program = credix_client::ID,
+        constraint = credix_lp_depository_0_pass.user == credix_lp_depository_0.key() @UxdError::InvalidCredixPass,
+        constraint = credix_lp_depository_0_pass.disable_withdrawal_fee @UxdError::InvalidCredixPassNoFees,
+    )]
+    pub credix_lp_depository_0_pass: Account<'info, credix_client::CredixPass>,
+
+    /// #19
+    pub credix_lp_depository_0_global_market_state:
+        Box<Account<'info, credix_client::GlobalMarketState>>,
+
+    /// #20 - CHECK: unused by us, checked by credix
+    pub credix_lp_depository_0_signing_authority: AccountInfo<'info>,
+
+    /// #21
+    #[account(mut)]
+    pub credix_lp_depository_0_liquidity_collateral: Box<Account<'info, TokenAccount>>,
+
+    /// #22
+    #[account(mut)]
+    pub credix_lp_depository_0_shares_mint: Box<Account<'info, Mint>>,
+
+    /// #23
+    pub system_program: Program<'info, System>,
+
+    /// #24
+    pub token_program: Program<'info, Token>,
+
+    /// #25
+    pub associated_token_program: Program<'info, AssociatedToken>,
+
+    /// #26
+    pub mercurial_vault_program: Program<'info, mercurial_vault::program::Vault>,
+
+    /// #27
+    pub credix_program: Program<'info, credix_client::program::Credix>,
+
+    /// #28
+    pub uxd_program: Program<'info, crate::program::Uxd>,
+
+    /// #29
+    pub rent: Sysvar<'info, Rent>,
+}
+
+struct DepositoryInfoForMintWithRouter {
+    pub weight_bps: u16,
+    pub redeemable_amount_under_management: u128,
+    pub redeemable_amount_under_management_cap: u128,
+}
+
+pub(crate) fn handler(ctx: Context<MintWithRouter>, collateral_amount: u64) -> Result<()> {
+    // Gather all the onchain states we need (caps, weights and supplies)
+    let controller = ctx.accounts.controller.load()?;
+    let identity_depository = ctx.accounts.identity_depository.load()?;
+    let mercurial_vault_depository_0 = ctx.accounts.mercurial_vault_depository_0.load()?;
+    let credix_lp_depository_0 = ctx.accounts.credix_lp_depository_0.load()?;
+
+    let redeemable_circulating_supply_before = controller.redeemable_circulating_supply;
+    let redeemable_circulating_supply_after = redeemable_circulating_supply_before
```

Comment:
This is not always true. Depositories may have minting fees, resulting in less than 1:1 UXD minted.

Reply:
True, this was meant more as an estimation, otherwise the maths are massively more complex. not supposed to be 100% accurate.

Overshooting the targets a little bit will ensure everything fits in the sum of all depositories.

If there is any discrepencies, the next mint will compensate for it, since the redeemable_under_management will be modified according to the mint fees.

Maybe I should rename this to `estimated_circulating_supply` ?

Reply:
Fair enough, done!